cluster_mapping_rules.go

```diff
@@ -68,55 +63,50 @@ func NewClusterConfigChange(c Cluster) ClusterConfigChange {
 	}
 }
 
-// A Mapping is a set of rules mapping shards onto a storage cluster.
-type Mapping interface {
+// A ClusterMappingRuleProvider provides cluster mapping rules
+type ClusterMappingRuleProvider interface {
 	// RulesForShard returns the mapping rules that currently apply to the given shard, plus
 	// a channel that can be used to receive new mapping rules
-	RulesForShard(n int) ([]MappingRule, <-chan []MappingRule, error)
-
-	// Clusters returns all of the currently defined clusters, plus a channel
-	// that can be used to receive cluster changes (new clusters, changes to
-	// cluster configurations)
-	Clusters() ([]Cluster, <-chan ClusterConfigChange, error)
+	RulesForShard(n int) ([]ClusterMappingRule, <-chan []ClusterMappingRule, error)
```

Comment:
Yeah it's not great, maybe we should wrap up the channel:
```go
type Events interface {
  OnUpdate(func (value interface{}))
  Close()
}

type EventsEmitter interface {
  Events
  Emit(value interface{})
}

type emitter struct {
  valueCh chan interface{}
  quitCh chan struct{}
}

func NewEventsEmitter(onCloseFn func(emitter EventsEmitter)) EventsEmitter {
  e := &emitter{make(chan interface{}), make(chan struct{})}
  if onCloseFn != nil {
    go func() {
     <-e.quitCh
     onCloseFn(e)
    }()
  }
  return e
}

func (e *emitter) Emit(value interface{}) {
  e.valueCh <- value
}

func (e *emitter) On(fn func(value interface{})) {
  go func() {
    for value := range e.valueCh {
      // Don't block emitters
      go func() { fn(value) }()
    }
  }()
}

func (e *emitter) Close() {
  e.quitCh <- struct{}{}
  close(e.valueCh)
  close(e.quitCh)
}
```

Then for a concrete implementation:
```go
type ClusterMappingRuleUpdates interface {
  OnUpdate(func (value ClusterMappingRule))
  Close()
}

type rulesEventEmitter struct {
  EventsEmitter
}

func newRulesEventEmitter(onCloseFn func(e *rulesEventEmitter)) *rulesEventEmitter {
  return &rulesEventEmitter{NewEventsEmitter(func (e EventsEmitter) { 
    onCloseFn(e.(*rulesEventEmitter))
  }
}

func (e *rulesEventEmitter) emit(value ClusterMappingRule) {
  e.EventsEmitter.Emit(value)
}

func (e *rulesEventEmitter) OnUpdate(fn func (value ClusterMappingRule)) {
  e.EventsEmitter.OnUpdate(func (value interface{}) { 
    fn(value.(ClusterMappingRule))
  })
}

func (e *rulesEventEmitter) Close() {
  e.EventsEmitter.Close()
}
```

In use:

```go
func (p ClusterMappingRuleProvider) Rules() ([]ClusterMappingRule, ClusterMappingRuleEvents, error) {
  emitter := newRulesEventEmitter(func(e *rulesEventEmitter) {
    // delete "e" from s.emitters
  }
  s.emitters = append(s.emitters, emitter)
  return p.rules, emitter, nil
}

// To enqueue simply for over s.listeners and emit(...) to each
```